src/main/java/com/uid2/operator/service/EncryptedTokenEncoder.java

```diff
@@ -144,17 +145,38 @@ private RefreshToken decodeRefreshTokenV3(Buffer b, byte[] bytes) {
     }
 
     @Override
-    public AdvertisingToken decodeAdvertisingToken(String s) {
-        final byte[] bytes = EncodingUtils.fromBase64(s);
-        final Buffer b = Buffer.buffer(bytes);
+    public AdvertisingToken decodeAdvertisingToken(String base64AdvertisingToken) {
+        if (base64AdvertisingToken.length() < 4) {
+            throw new IllegalArgumentException("Advertising token is too short");
+        }
 
-        if (b.getByte(1) == TokenVersion.V3.rawVersion) {
-            return decodeAdvertisingTokenV3(b, bytes);
-        } else if (b.getByte(0) == TokenVersion.V2.rawVersion) {
+        String headerStr = base64AdvertisingToken.substring(0, 4);
+        boolean isBase64UrlEncoding = (headerStr.indexOf('-') != -1 || headerStr.indexOf('_') != -1);
+        byte[] headerBytes = isBase64UrlEncoding ? Uid2Base64UrlCoder.decode(headerStr) : Base64.getDecoder().decode(headerStr);
+
+        if (headerBytes[0] == TokenVersion.V2.rawVersion) {
+            final byte[] bytes = EncodingUtils.fromBase64(base64AdvertisingToken);
+            final Buffer b = Buffer.buffer(bytes);
             return decodeAdvertisingTokenV2(b);
+        }
+
+        //Java's byte is signed, so we convert to unsigned before checking the enum
+        int unsignedByte = ((int) headerBytes[1]) & 0xff;
+
+        byte[] bytes;
+        TokenVersion tokenVersion;
+        if (unsignedByte == TokenVersion.V3.rawVersion) {
+            bytes = EncodingUtils.fromBase64(base64AdvertisingToken);
+            tokenVersion = TokenVersion.V3;
+        } else if (unsignedByte == TokenVersion.V4.rawVersion) {
+            bytes = Uid2Base64UrlCoder.decode(base64AdvertisingToken);  //same as V3 but use Base64URL encoding
+            tokenVersion = TokenVersion.V4;
         } else {
             throw new IllegalArgumentException("Invalid advertising token version");
         }
+
+        final Buffer b = Buffer.buffer(bytes);
```

Comment:
Nit: `Buffer.buffer(byte[])` copies the bytes. Since we own the `byte[]`, we can avoid the copying with:

```suggestion
        final Buffer b = Buffer.buffer(Unpooled.wrappedBuffer(bytes));
```

Reply:
as above, we can leave this as part of UID2-1206
(noting this is for /token/validate which is rarely called)